src/appleseed/renderer/modeling/aov/cryptomatte.cpp

```diff
@@ -0,0 +1,908 @@
+                }
+            }
+            return;
+        }
+    };
+
+    const char* CryptomatteObjectAOVModel = "cryptomatte_object_aov";
+    const char* CryptomatteMaterialAOVModel = "cryptomatte_material_aov";
+}
+
+
+//
+// CryptomatteAOV class implementation.
+//
+
+struct CryptomatteAOV::Impl
+{
+    vector<WeightMap>                   m_pixel_samples;
+    map<pair<int, int>, NameMap>        m_tile_name_map;
+    unique_ptr<Image>                   m_image;
+    size_t                              m_num_layers;
+    CryptomatteAOV::CryptomatteType     m_layer_type;
+};
+
+CryptomatteAOV::CryptomatteAOV(const ParamArray& params)
+  : AOV("cryptomatte", params)
+  , impl(new Impl())
+{
+    const string cryptomatte_type = params.get_optional<string>("cryptomatte_type", "object_names", make_vector("object_names" , "material_names"));
+    
+    if (cryptomatte_type == "object_names")
+    {
+        impl->m_layer_type = CryptomatteType::ObjectNames;
+        set_name(CryptomatteObjectAOVModel);
+    }
+    else if (cryptomatte_type == "material_names")
+    {
+        impl->m_layer_type = CryptomatteType::MaterialNames;
+        set_name(CryptomatteMaterialAOVModel);
+    }
+
+    impl->m_num_layers = params.get_optional<size_t>("cryptomatte_num_layers", 6);
+}
+
+CryptomatteAOV::~CryptomatteAOV()
+{
+    delete impl;
+}
+
+const char* CryptomatteAOV::get_model() const
+{
+    if (impl->m_layer_type == CryptomatteAOV::CryptomatteType::ObjectNames)
+        return CryptomatteObjectAOVModel;
+    else
+        return CryptomatteMaterialAOVModel;
+}
+
+size_t CryptomatteAOV::get_channel_count() const
+{
+    return 0;
+}
+
+const char** CryptomatteAOV::get_channel_names() const
+{
+    return nullptr;
+}
+
+bool CryptomatteAOV::has_color_data() const
+{
+    return false;
+}
+
+void CryptomatteAOV::create_image(
+    const size_t    canvas_width,
+    const size_t    canvas_height,
+    const size_t    tile_width,
+    const size_t    tile_height,
+    ImageStack&     aov_images)
+{
+    // Create underlying images.
+    const size_t num_channels = (impl->m_num_layers * 2) + 3;
+    impl->m_image.reset(
+        new Image(
+            canvas_width,
+            canvas_height,
+            tile_width,
+            tile_height,
+            num_channels,
+            PixelFormatFloat));
+    impl->m_pixel_samples.resize(canvas_width * canvas_height, WeightMap(impl->m_num_layers));
+    clear_image();
+}
+
+void CryptomatteAOV::clear_image()
+{
+    const size_t num_channels = (impl->m_num_layers * 2) + 3;
+    vector<float> pixel_values(num_channels, 0.0f);
+    const auto image_props = impl->m_image->properties();
+    for (size_t ry = 0, ey = image_props.m_canvas_height; ry < ey; ++ry)
+    {
+        for (size_t rx = 0, ex = image_props.m_canvas_width; rx < ex; ++rx)
+        {
+                impl->m_image->set_pixel(rx, ry, pixel_values.data());
```

Comment:
Fix indentation.

Reply:
That's going to be very slow because for each pixel `ICanvas::set_pixel()` needs to find the associated tile.

Here's a faster implementation (untested, please check):

```cpp
for (size_t ty = 0, ey = image_props.m_tile_count_y; ty < ey; ++ty)
{
    for (size_t tx = 0, ex = image_props.m_tile_count_x; tx < ex; ++tx)
    {
        Tile& tile = impl->m_image->tile(tx, ty);
        memset(tile.get_storage(), 0, tile.get_size());
    }
}
```

That works because `0` is the bit pattern for `0.0f` (how convenient!).